File: adc02.ml
let file = "test02.txt"
(* let file = "input02.txt" *)

let process_bag s =
  let r = ref 0 and g = ref 0 and b = ref 0 in
  let colors = String.split_on_char ',' s in
  List.iter
    (fun s ->
      match String.split_on_char ' ' s with
      | [ ""; i; "red" ] -> r := int_of_string i
      | [ ""; i; "green" ] -> g := int_of_string i
      | [ ""; i; "blue" ] -> b := int_of_string i
      | _ -> failwith s)
    colors;
  (!r, !g, !b)

let process_bags s =
  let bag_list = String.split_on_char ';' s in
  List.map process_bag bag_list

let process_line s =
  match String.split_on_char ':' s with
  | [ header; bags ] ->
      let id =
        String.sub header 5 (String.length header - 5)
        |> int_of_string
      in
      (id, process_bags bags)
  | _ -> failwith "No game id"

let is_possible =
  List.for_all (fun (r, g, b) -> r <= 12 && g <= 13 && b <= 14)

let _ =
  let data = open_in file in
  let sum1 = ref 0 in
  try
    while true do
      let line = input_line data in
      let id, bags = process_line line in
      if is_possible bags then sum1 := !sum1 + id
    done
  with End_of_file ->
    close_in data;
    Printf.printf "Part one: %d\n" !sum1

File: adc02/adc02.py
from collections import defaultdict


def process_bag(str):
    d = defaultdict(int)
    for colored_cubes in str.split(", "):
        [num_str, color] = colored_cubes.split(" ")
        assert color not in d
        d[color] = int(num_str)
    return d


def process_line(str):
    [header, bags_str] = str.split(": ")
    id = int(header.split(" ")[1])
    bags = [process_bag(bag) for bag in bags_str.split("; ")]
    return id, bags


def is_possible(bags):
    for bag in bags:
        if bag["red"] > 12 or bag["green"] > 13 or bag["blue"] > 14:
            return False
    return True


def power(bag):
    d = defaultdict(int)
    for bag in bags:
        for color in bag:
            d[color] = max(d[color], bag[color])
    return d["red"] * d["green"] * d["blue"]


with open("input02.txt", "r") as file:
    sum1 = 0
    sum2 = 0
    for line in file:
        # Pesky endline
        id, bags = process_line(line[:-1])
        if is_possible(bags):
            sum1 += id
        sum2 += power(bags)
    print(f"Part one: {sum1}")
    print(f"Part two: {sum2}")
